### PR TITLE
Support video bitrate selection

### LINKED ISSUE #24: Add transcoding selector
Support playback quality controls in native player to limit bitrate (and resolution) by transcoding the video files.

### PR DESCRIPTION
**Pull request by Maxr1998**
Add support for bitrate selection and improves transcoding playback.

Supersedes and closes #274.

Closes #24.

### CODE REVIEW

**Review by roeii — DISMISSED** [review]
Looks good so far. What work is missing to make this complete?

**Comment by nielsvanvelzen** [comment]
@Maxr1998 what's the state of this PR?

**Comment by davidngrc** [comment]
This feature is awesome, why not yet merge?
the HTML5 player in my android app or web don't work on mobile network on tvheadend with apache, somehow the ExoPlayer works, but just missing transcode, please help implement this. Thank you.

**Comment by Maxr1998** [comment]
Let's get this in, I'd say. It doesn't yet respect the bitrate limits from client settings, that is tracked in #564 and will be an upcoming PR.

File: app/src/main/res/drawable/ic_settings_white_24dp.xml [hunk 1]
```diff
@@ -0,0 +1,5 @@
...
...-0.19,0.15 -0.24,0.42 -0.12,0.64l2,3.46c0.09,0.16 0.26,0.25 0.44,0.25 0.06,0 0.12,-0.01 0.17,-0.03l2.49,-1c0.52,0.4 1.08,0.73 1.69,0.98l0.38,2.65c0.03,0.24 0.24,0.42 0.49,0.42h4c0.25,0 0.46,-0.18 0.49,-0.42l0.38,-2.65c0.61,-0.25 1.17,-0.59 1.69,-0.98l2.49,1c0.06,0.02 0.12,0.03 0.18,0.03 0.17,0 0.34,-0.09 0.43,-0.25l2,-3.46c0.12,-0.22 0.07,-0.49 -0.12,-0.64l-2.11,-1.65zM17.45,11.27c0.04,0.31 0.05,0.52 0.05,0.73 0,0.21 -0.02,0.43 -0.05,0.73l-0.14,1.13 0.89,0.7 1.08,0.84 -0.7,1.21 -1.27,-0.51 -1.04,-0.42 -0.9,0.68c-0.43,0.32 -0.84,0.56 -1.25,0.73l-1.06,0.43 -0.16,1.13 -0.2,1.35h-1.4l-0.19,-1.35 -0.16,-1.13 -1.06,-0.43c-0.43,-0.18 -0.83,-0.41 -1.23,-0.71l-0.91,-0.7 -1.06,0.43 -1.27,0.51 -0.7,-1.21 1.08,-0.84 0.89,-0.7 -0.14,-1.13c-0.03,-0.31 -0.05,-0.54 -0.05,-0.74s0.02,-0.43 0.05,-0.73l0.14,-1.13 -0.89,-0.7 -1.08,-0.84 0.7,-1.21 1.27,0.51 1.04,0.42 0.9,-0.68c0.43,-0.32 0.84,-0.56 1.25,-0.73l1.06,-0.43 0.16,-1.13 0.2,-1.35h1.39l0.19,1.35 0.16,1.13 1.06,0.43c0.43,0.18 0.83,0.41 1.23,0.71l0.91,0.7 1.06,-0.43 1.27,-0.51 0.7,1.21 -1.07,0.85 -0.89,0.7 0.14,1.13zM12,8c-2.21,0 -4,1.79 -4,4s1.79,4 4,4 4,-1.79 4,-4 -1.79,-4 -4,-4zM12,14c-1.1,0 -2,-0.9 -2,-2s0.9,-2 2,-2 2,0.9 2,2 -0.9,2 -2,2z"/>
```
Comment (nielsvanvelzen) [inline app/src/main/res/drawable/ic_settings_white_24dp.xml:4]: I think you forgot to format, end of the tag doesn't have a space

File: app/src/main/java/org/jellyfin/mobile/player/ui/PlayerMenus.kt [hunk 2]
```diff
@@ -225,10 +263,23 @@ class PlayerMenus(
         fragment.suppressControllerAutoHide(false)
     }
 
+    private fun formatBitrate(bitrate: Double): String {
+        val (value, unit) = when {
+            bitrate > BITRATE_MEGA_BIT -> bitrate / BITRATE_MEGA_BIT to " Mbps"
+            bitrate > BITRATE_KILO_BIT -> bitrate / BITRATE_KILO_BIT to " kbps"
+            else -> bitrate to " bps"
```
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/mobile/player/ui/PlayerMenus.kt:270]: Do we want to use (untranslateable-) strings for these?

File: app/src/main/java/org/jellyfin/mobile/player/ui/PlayerMenus.kt [hunk 3]
```diff
@@ -217,6 +243,18 @@ class PlayerMenus(
         subtitlesButton.setImageState(stateSet, true)
     }
 
+    private fun buildQualityMenu(menu: Menu, videoWidth: Int, videoHeight: Int) {
+        menu.clear()
+        val options = qualityOptionsProvider.getApplicableQualityOptions(videoWidth, videoHeight)
+        options.map { option ->
+            val title = when (val bitrate = option.bitrate) {
+                0 -> context.getString(R.string.menu_item_auto)
+                else -> "${option.maxHeight}p - ${formatBitrate(bitrate.toDouble())}"
```
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/mobile/player/ui/PlayerMenus.kt:252]: Do we want to use (untranslateable-) strings for these?

File: app/src/main/java/org/jellyfin/mobile/player/qualityoptions/QualityOptionsProvider.kt [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        QualityOption(maxHeight = 360, bitrate = 420000),
+        QualityOption(maxHeight = 0, bitrate = 0), // auto
+    )
+
+    @Suppress("MagicNumber")
+    fun getApplicableQualityOptions(videoWidth: Int, videoHeight: Int): List<QualityOption> {
+        // If the aspect ratio is less than 16/9 (1.77), set the width as if it were pillarboxed
+        // i.e. 4:3 1440x1080 -> 1920x1080
+        val maxAllowedWidth = when {
+            videoWidth / videoHeight < 16 / 9 -> videoHeight * (16 / 9)
```
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/mobile/player/qualityoptions/QualityOptionsProvider.kt:28]: Probably makes sense to move `16/9` to a constant so it's pre-calculated on use (ASPECT_RATIO_16_9)

**Review by nielsvanvelzen — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/jellyfin/mobile/player/ui/PlayerMenus.kt (hunk 3 above)
Comment (Maxr1998) [reply app/src/main/java/org/jellyfin/mobile/player/ui/PlayerMenus.kt:252]: Don't think it's necessary.

File: app/src/main/java/org/jellyfin/mobile/player/ui/PlayerMenus.kt (hunk 2 above)
Comment (Maxr1998) [reply app/src/main/java/org/jellyfin/mobile/player/ui/PlayerMenus.kt:270]: Don't think it's necessary.

File: app/src/main/java/org/jellyfin/mobile/player/qualityoptions/QualityOptionsProvider.kt (hunk 4 above)
Comment (Maxr1998) [reply app/src/main/java/org/jellyfin/mobile/player/qualityoptions/QualityOptionsProvider.kt]: The code was actually broken anyway, all of those values were `Ints`…
I fixed it and use a `Rational` to compare them now.

File: app/src/main/res/drawable/ic_settings_white_24dp.xml (hunk 1 above)
Comment (Maxr1998) [reply app/src/main/res/drawable/ic_settings_white_24dp.xml]: Done.

**Review by nielsvanvelzen — APPROVED** [review]

**Pull request merged**